File: src/ch_core/transaction_manager.cpp
#include "src/ch_core/transaction_manager.h"

namespace ch {

TransactionManager::TransactionManager(std::size_t max_depth)
    : max_depth_(max_depth) {}

void TransactionManager::record_command(const CommandRecord& cmd) {
    if (cmd.payloads.empty()) {
        return;
    }

    if (active_batch_.has_value()) {
        for (const auto& payload : cmd.payloads) {
            active_batch_->payloads.push_back(payload);
        }
        return;
    }

    undo_stack_.push_back(cmd);
    redo_stack_.clear();

    if (undo_stack_.size() > max_depth_) {
        undo_stack_.erase(undo_stack_.begin());
    }
}

void TransactionManager::begin_transaction(const std::string& description) {
    if (active_batch_.has_value()) {
        commit_transaction();
    }
    CommandRecord record;
    record.description = description;
    record.action_type = TransactionActionType::batch_transaction;
    active_batch_ = record;
}

void TransactionManager::commit_transaction() {
    if (!active_batch_.has_value()) {
        return;
    }

    CommandRecord batch = std::move(*active_batch_);
    active_batch_.reset();

    if (!batch.payloads.empty()) {
        undo_stack_.push_back(batch);
        redo_stack_.clear();

        if (undo_stack_.size() > max_depth_) {
            undo_stack_.erase(undo_stack_.begin());
        }
    }
}

void TransactionManager::cancel_transaction() {
    active_batch_.reset();
}

bool TransactionManager::can_undo() const {
    return !undo_stack_.empty() && !active_batch_.has_value();
}

bool TransactionManager::can_redo() const {
    return !redo_stack_.empty() && !active_batch_.has_value();
}

std::optional<CommandRecord> TransactionManager::undo() {
    if (!can_undo()) {
        return std::nullopt;
    }

    CommandRecord cmd = std::move(undo_stack_.back());
    undo_stack_.pop_back();

    redo_stack_.push_back(cmd);
    return cmd;
}

std::optional<CommandRecord> TransactionManager::redo() {
    if (!can_redo()) {
        return std::nullopt;
    }

    CommandRecord cmd = std::move(redo_stack_.back());
    redo_stack_.pop_back();

    undo_stack_.push_back(cmd);
    return cmd;
}

void TransactionManager::clear_history() {
    undo_stack_.clear();
    redo_stack_.clear();
    active_batch_.reset();
}

} // namespace ch

File: tools/map_forge/core/transaction_manager.py
"""
Transaction & Undo/Redo Engine for Map Forge (CH_TRANSACTION_V1).
Manages reversible mutation commands and atomic batch transactions over MapModel.
"""

from typing import Dict, Any, List, Optional
from tools.map_forge.core.map_model import MapModel


class MapTransactionManager:
    def __init__(self, map_model: MapModel, max_depth: int = 100):
        self.map_model = map_model
        self.max_depth = max_depth
        self.undo_stack: List[Dict[str, Any]] = []
        self.redo_stack: List[Dict[str, Any]] = []
        self.active_batch: Optional[Dict[str, Any]] = None

    def record_mutation(self, action_type: str, description: str, inverse_changes: List[Dict[str, Any]], forward_changes: List[Dict[str, Any]]):
        if not inverse_changes or not forward_changes:
            return

        cmd = {
            "action_type": action_type,
            "description": description,
            "inverse_changes": inverse_changes,
            "forward_changes": forward_changes
        }

        if self.active_batch is not None:
            self.active_batch["inverse_changes"] = inverse_changes + self.active_batch["inverse_changes"]
            self.active_batch["forward_changes"] = self.active_batch["forward_changes"] + forward_changes
            return

        self.undo_stack.append(cmd)
        self.redo_stack.clear()

        if len(self.undo_stack) > self.max_depth:
            self.undo_stack.pop(0)

    def begin_transaction(self, description: str = "Batch Operation"):
        if self.active_batch is not None:
            self.commit_transaction()

        self.active_batch = {
            "action_type": "batch_transaction",
            "description": description,
            "inverse_changes": [],
            "forward_changes": []
        }

    def commit_transaction(self):
        if self.active_batch is None:
            return

        batch = self.active_batch
        self.active_batch = None

        if batch["forward_changes"]:
            self.undo_stack.append(batch)
            self.redo_stack.clear()

            if len(self.undo_stack) > self.max_depth:
                self.undo_stack.pop(0)

    def cancel_transaction(self):
        if self.active_batch is None:
            return
        
        # Rollback transient batch changes applied so far
        for change in self.active_batch["inverse_changes"]:
            self._apply_change(change)
        
        self.active_batch = None

    def can_undo(self) -> bool:
        return len(self.undo_stack) > 0 and self.active_batch is None

    def can_redo(self) -> bool:
        return len(self.redo_stack) > 0 and self.active_batch is None

    def undo(self) -> Optional[Dict[str, Any]]:
        if not self.can_undo():
            return None

        cmd = self.undo_stack.pop()
        for change in cmd["inverse_changes"]:
            self._apply_change(change)

        self.redo_stack.append(cmd)
        return cmd

    def redo(self) -> Optional[Dict[str, Any]]:
        if not self.can_redo():
            return None

        cmd = self.redo_stack.pop()
        for change in cmd["forward_changes"]:
            self._apply_change(change)

        self.undo_stack.append(cmd)
        return cmd

    def clear_history(self):
        self.undo_stack.clear()
        self.redo_stack.clear()
        self.active_batch = None

    def _apply_change(self, change: Dict[str, Any]):
        type_ = change.get("type")
        if type_ == "terrain":
            self.map_model.set_terrain(change["x"], change["y"], change["texture"],
                                       change.get("terrain_definition"), change.get("apply_semantics", False))
        elif type_ == "add_building":
            self.map_model.add_building(
                change["definition_id"],
                change["x"],
                change["y"],
                change.get("rotation", 0),
                instance_id=change.get("instance_id")
            )
        elif type_ == "remove_building":
            self.map_model.remove_building_at(change["x"], change["y"])
        elif type_ == "set_road":
            self.map_model.set_road(change["x"], change["y"], change["present"])
